go/vt/vtgate/vcursor_impl.go

```diff
@@ -138,19 +138,36 @@ func (vc *vcursorImpl) TargetString() string {
 	return vc.safeSession.TargetString
 }
 
-// Execute performs a V3 level execution of the query.
-func (vc *vcursorImpl) Execute(method string, query string, BindVars map[string]*querypb.BindVariable, isDML bool) (*sqltypes.Result, error) {
-	qr, err := vc.executor.Execute(vc.ctx, method, vc.safeSession, vc.marginComments.Leading+query+vc.marginComments.Trailing, BindVars)
-	if err == nil {
+// Execute is part of the engine.VCursor interface.
+func (vc *vcursorImpl) Execute(method string, query string, bindVars map[string]*querypb.BindVariable, isDML bool) (*sqltypes.Result, error) {
+	return vc.executeByOrder(method, query, bindVars, isDML, commitOrderNormal)
+}
+
+// ExecutePre is part of the engine.VCursor interface.
+func (vc *vcursorImpl) ExecutePre(method string, query string, bindVars map[string]*querypb.BindVariable, isDML bool) (*sqltypes.Result, error) {
+	return vc.executeByOrder(method, query, bindVars, isDML, commitOrderPre)
+}
+
+// ExecutePost is part of the engine.VCursor interface.
+func (vc *vcursorImpl) ExecutePost(method string, query string, bindVars map[string]*querypb.BindVariable, isDML bool) (*sqltypes.Result, error) {
+	return vc.executeByOrder(method, query, bindVars, isDML, commitOrderPost)
+}
+
+func (vc *vcursorImpl) executeByOrder(method string, query string, bindVars map[string]*querypb.BindVariable, isDML bool, co commitOrder) (*sqltypes.Result, error) {
```

Comment:
I went this route because we otherwise introduce a circular dependency on `commitOrder`: vtgate calls engine, which calls into vindexes, and the vindex calls back into VCursor (as interface).

We could move a `commitOrder` to vindexes and export it. Do you think that will still remain readable?

Or, I just thought of this: define it in the proto. I think proto looks like the best place.